Subtitle sync

mcosmin222 opened the pull request:
This adds subtitle syncing. I.e make subtitles appear later or earlier in the media timeline.

lukasf commented:
Let's continue discussion here, so we see changes without switching form issue to pr and back.

It would be good if the cue moving stuff is handled inside the SubtitleProvider class. The FFmpegInteropMSS class is huge already, let's not blow that up with subtitle-specific implementations. Moving it into SubtitleProvider has the benefit that we can use the timer, mutex and tick function that are already defined there.

I currently don't think that we need the AutoResetEvent. If we lock the mutex in both the set method and the tick method, then nothing can go wrong. Even if the previous tick function comes when the next set was already triggered, it would use the updated values already, since Tick can only execute after next set is completed (mutex locked). So I think there cannot be a situation where we get incorrect results. The same is already true for the anti-flicker solution, which is also safe even if Tick method runs async.

About removed internal subs: Internal subs are expected to be in sync with the video, so moving them more than a tiny bit will be a rare situation. Bigger delays will usually only be required for external subs. Still, we could decide to only remove bitmap subs on CueExited but keep all text subs.

Let's also keep in mind that we will always have problems with bigger negtive delay on internal subs. When seek is used, we won't get subs that are stored before the target frame. So subs would only come after some seconds of playback. We cannot do much about this. It also happens with other players.

mcosmin222 commented:
> It would be good if the cue moving stuff is handled inside the SubtitleProvider class. The FFmpegInteropMSS class is huge already, let's not blow that up with subtitle-specific implementations. Moving it into SubtitleProvider has the benefit that we can use the timer, mutex and tick function that are already defined there.

Agreed. I will move it over there.

>
I currently don't think that we need the AutoResetEvent. If we lock the mutex in both the set method and the tick method, then nothing can go wrong. Even if the previous tick function comes when the next set was already triggered, it would use the updated values already, since Tick can only execute after next set is completed (mutex locked). So I think there cannot be a situation where we get incorrect results. The same is already true for the anti-flicker solution, which is also safe even if Tick method runs async.

Ok, we could give it a try. Sounds reasonable.

> About removed internal subs: Internal subs are expected to be in sync with the video, so moving them more than a tiny bit will be a rare situation. Bigger delays will usually only be required for external subs. Still, we could decide to only remove bitmap subs on CueExited but keep all text subs.
> 
> Let's also keep in mind that we will always have problems with bigger negtive delay on internal subs. When seek is used, we won't get subs that are stored before the target frame. So subs would only come after some seconds of playback. We cannot do much about this. It also happens with other players.

Well, even if we do remove the cues, it would still make little sense to delay the embedded ones by any large amount. Right now I delay by seconds, just to be sure it works. So I tested that...

...I don't think we need to remove the cues. It doesn't make much sense. So lets say we have a cue which is 10 seconds after the current time, even if we hasten the cues by 10 seconds (which would normally bring it into view), it won't work, since we don't have that cue yet decoded. For embedded we would need to seek the subtitle stream individually, and I don't think that is possible without seeking the entire input stream.

So yah, for embedded subtitles I say we just add the offset to start position calculus and that is it. No need to remove the cue removal ^^

lukasf commented:
Yes maybe we should not even touch removal. Things will repair when playing for a few seconds and the same problem will occur anyways after each seek.

> EDIT: actually no, for external subs, we no longer have the samples provider. For embedded subs, yes, the calculus will be done there. I guess we can introduce a new class, "ExternalSubtitleProvider" and move the logic to it.

For external audio, I move the actual real sample provider instance from the external parser MSS into the main MSS instance. ;) This has a few side effects to handle, but it worked well. But for external subs it might be better to have a specific implementation, as you proposed.

lukasf commented:
For start time of new subs (live playback of embedded subs), we just need to add the subtitle delay in SubtitleProvider.h, where m_startOffset is already added:

    virtual void QueuePacket(AVPacket *packet) override
    {
    ...
				position.Duration = LONGLONG(av_q2d(m_pAvStream->time_base) * 10000000 * packet->pts) - m_startOffset;

mcosmin222 commented:
Yes, that's how I did it in my test. I will add the new class and push the code.

mcosmin222 commented:
I still think there is a possibility of thread races, and should probably be synchronised against the main mutex. What happens if mid way through parsing an external sub, someone changes the delay? This will result in some cues having different offsets.

lukasf commented:
I already wanted to point that out: It would be better if we do not access the config class property here. Instead, during creation of the original MSS we should get the value once from the config class, store it and use it for all subtitles. If we now add external subs, use the stored value in the MSS. Only when the set method is called on the MSS, we get this new value and update all subs. That way, all changes to the delay are mutex locked. 

This also allows users to change the value in the config class, to have a default delay (something you currently prevent, but should be possible). So now the property could be get/set.

mcosmin222 commented:
I will give it a try in the evening.

mcosmin222 commented:
I still don't think that would be enough. The actual parsing of the external subtitle is done outside of mutex lock. If we do mutex lock it, it might interrupt playback.

lukasf commented:
You are right. I would have proposed to just ignore that scenario, but the changeset you just pushed looks good.

mcosmin222 commented:
What about syncing the external subtitle providers with that same mutex? I would personally do it, makes it much easier to consume the API.

lukasf commented:
I am not sure I understand what you mean. If you think it is a good idea, just try it. :)

mcosmin222 commented:
So while I was enjoying lunch i had an idea: should we add offsets to Duration of cues as well? I don't know of any player doing this, but I imagine it might be some sort of ease of access feature.

lukasf commented:
I had this very same idea quite some time ago! Indeed no player supports it right now, but I would personally enable it if available.

It raises some questions on whether to allow overlapping of subs or not. It would be good if we could add it in a way so no overlapping is added. So only extend the sub until the point where the next sub comes. Of course if original subs do already overlap, we can ignore that.

The prevent overlap part is probably a bit difficult to realize, since the next sub might not yet be parsed (for embedded subs), and we alsoi need to be able to undo modifications when duration is reduced - which would force us to remember the original duration.

mcosmin222 commented:
I think we should allow overlapping. I don't see the problem if they did overlap. I don't think anyone would increase durations by more than a few seconds.

lukasf commented:
When people talk fast (e.g. discussion), subs usually come right after each other. If we add a delay, we would have subs moving around through the whole discussion. This does not look nice, and does not improve readability of subs (which is probably the goal of having extended duration).

mcosmin222 commented:
> 
> 
> What about syncing the external subtitle providers with that same mutex? I would personally do it, makes it much easier to consume the API.

Well it was a bad idea. 



> 
> 
> When people talk fast (e.g. discussion), subs usually come right after each other. If we add a duration, we would have subs moving around through the whole discussion. This does not look nice, and does not improve readability of subs (which is probably the goal of having extended duration).
> 
> We could also add durations as simple solution first (with overlapping) and add overlap prevention later, as optional feature.

Indeed that could be a problem. But if we don't have all subtitles longer, then the goal of making them easier to read also cannot be accomplished.

mcosmin222 commented:
Anyway, anything i've missed?

lukasf commented:
Enhanced subtitle duration without overlap prevention: Makes many subs easier to read but some subs harder to read.

Enhanced subtitle duration with overlap prevention: Makes many subs easier to read but some stay as they are.

So for me, I would use the feature, but only with overlap prevention.

mcosmin222 commented:
Ok then. We should probably leave that for another PR.

mcosmin222 commented:
@lukasf do you think this is ready for the merge?^^

lukasf commented on FFmpegInterop/AutoResetEvent.h:
```diff
@@ -0,0 +1,43 @@
+#include <thread>
+#include <stdio.h>
+
+class AutoResetEvent
```
I think this is not needed anymore right?

mcosmin222 replied:
Yea, I was thinking about that. Maybe we will need it in the future. But I guess we can add it then.

lukasf commented on FFmpegInterop/ExternalSubtitleProvider.h:
```diff
@@ -0,0 +1,110 @@
+		{
+			mutex.lock();
+			this->newOffset = newOffset;
+			this->oldOffset = currentOffset;
```
I think here is a potential race condition. If set method is called twice before tick comes, then the calculations in OnTick will be wrong (because oldDelay is not correct anymore). It could be fixed like this: Have two fields currentDelay and newDelay. Initialize currentDelay in the constructor, with the value from the config class. In the set method, only set newDelay. Now in OnTick, you will always have the correct currentDelay, plus the latest newDelay.

lukasf commented on FFmpegInterop/ExternalSubtitleProvider.h:
```diff
@@ -0,0 +1,110 @@
+					//start time cannot be negative.
+					if (newStartPosition.Duration < 0)
+					{
+						newStartPosition.Duration = 0;
```
This gives me some headaches. If someone sets a bigger negative delay, and then sets back to normal, then the first cues will be all wrong. The only way to fix this is to have a list of cues that got speacial treatment here, with their unmodified negative durations. On next OnTick, we would have to check if the cue is in the list, and use the duration from there for all calculations.

lukasf commented on FFmpegInterop/FFmpegInteropMSS.h:
```diff
@@ -222,6 +229,7 @@ namespace FFmpegInterop
 
 	private:
 
+		DispatcherTimer^ timer;
```
This is not used, I think.

lukasf commented on FFmpegInterop/SubtitleProvider.h:
```diff
@@ -70,7 +72,7 @@ namespace FFmpegInterop
 				bool isDurationFixed = false;
 
-				position.Duration = LONGLONG(av_q2d(m_pAvStream->time_base) * 10000000 * packet->pts) - m_startOffset;
+				position.Duration = LONGLONG(av_q2d(m_pAvStream->time_base) * 10000000 * packet->pts) - m_startOffset + AdditionalStreamSampleOffset.Duration;
```
I know it would blow things up, but we could fix cues in the track here as well, like in the external subs provider. Maybe it is not really needed, not sure...

mcosmin222 commented:
Ok, I can agree with ...Delay.

And since most of the review is based on having thread races, maybe we should rethink how we are going to handle this. I think using too many variables and too many mutexes will make things harder to understand in the future, even for us. Unfortunately, C++ is really, really bad at handling multithreading.

I will try to think of a way to eliminate the possibility of thread races at root level (i.e the call to that SetSubtitlesDelay method in ffmpeginteropMSS) without crowding the main class itself. I am pretty sure we can continue to find such problems as we go cause this implementation is so prone to thread races.

And all of this because Microsoft forces us to use that DispatcherTimer.

mcosmin222 commented:
And by the way, I know I am bad at naming things xD

mcosmin222 commented:
So my idea is to return async actions wrapping the call to dispatcher timer. The actions would return when the subtitle is done updating its positions. Then the set subtitle delay in the ffmpeginterop mss would mutex lock and wait on all actions to complete, and then release the mutex. This will solve the source of all thread races at its root. Guess I will use the autoresetevent anyway, unless there is some other primitive i can use from c++/Cx.

As for the negative duration cues, I think your idea to save in a list should do the job..

lukasf commented:
Acutally, I don't think the threading is that bad. The only problem in the current code is that oldOffset is modified when the offset has not yet been applied. Move the change of oldOffset from the Set method to OnTick, and the race condition is gone. And I think we can also get rid of the other mutex (which locks adding external subs): Get the current offset at the start of AddExternalSubtitle and load the sub with that offset. When the sub is loaded (and main mutex is locked), we check if the current offset has changed. If it has changed, we now call SetOffset on the sub with the actual value. Now everything should be threadsafe, without any added locking or async methods.

The solution you propose would sure work. But it also adds quite some overhead, and it forces us to make the set method async.

One more benefit from my solution: It would allow us to add multiple external subs in one run (in parallel). With the current or your proposed solution, only one sub will be added at any time, due to the newly introduced mutex lock.

mcosmin222 commented:
Well, for loading multiple subs in parallel, I am not sure. If that happens we expose to other thread races... I.e changes in the config class. And the parsing of a subtitle is IO intensive, and i doubt someone would expect to do it in parallel. Most people would keep their media files on hard disk drives, and loading 2 in parallel would expose be slower than 2 one after the other.

The problem with moving the offset to on tick is that we cannot pass arguments to that event handler, so we will need some kind of global state, which means we have to lock anyway. 
And all of this is actually done on the UI thread anyway so it is not true parallelism. The overhead is minimal. In fact, I think updating each sub track 1 by 1 is probably the best way. Right now we flood the UI thread with several dispatcher timers. On slower CPUs this might cause freezes.

I will give some more thought about it.

lukasf commented:
I think my solution will be simple and thread safe. Maybe I did not describe it very well. I think I will just implement it (maybe on other branch), then we can discuss on that. Using OnTick is only really needed for a sub that is actually showing. All others can update directly.

mcosmin222 commented:
If you have time you can push it here and we can undo the conmit if it destroies the universe. 

If not, i will try it in the evening.

lukasf commented:
Ok I just pushed my changeset. The main difference for the race condition is setting the currentDelay only in OnTick, after it has really been applied. Now it does not matter how often Set method is called. OnTick will always have the actual current delay, and the latest new delay, so calculations are correct and output is right. 

I have also done the other changes I mentioned. Current delay is initialized in the constructor with the value that was used when creating the external MSS parser. If the delay has changed while we were parsing the stream (which is very unlikely), SetDelay is called again before the sub is added. So streams are always added with latest delay, even if delay changes inbetween.

mcosmin222 commented:
Ah I see what you mean now.
Yes, this make sense. But I still think there would be a race in AddExternalSubtitleAsync when we copy the config from the main instance to the subConfig. I say we should lock it...just that particular region.

mcosmin222 commented:
Actually, that might not be so much of a problem. Users should expect to have weird results when doing parallelism like that with subtitles and different encoding/properties for each file. It is a niche case anyway.

lukasf commented:
Even a lock would not help much, since we do not lock anything on config value changes. If people modify the config instance after it has been passed into the MSS constructor, while other async actions (or playback) are running, the behavior will always be a bit unpredictable.

mcosmin222 commented:
Ok cool. Thanks for that commit. I am now handling the merge and other review issues. I really didn't have time today until now.

mcosmin222 commented:
So I removed the autoresetevent class.
Renamed the "offset" to "delay"
Fixed the problem with negative duration cues. That one took some doing, sadly we cannot use either winRt Map or std::map to store the cue to duration pair, so I ended up using a vector<pair>

lukasf commented:
Cool, this looks good!

Yesterday I have been thinking if we can inherit ExternalSubtitleProvider from SubtitleProvider and move the delay stuff into SubtitleProvider. Then, it would automatically work for embedded subs as well, and in the MSS class we would not have to differentiate between embedded and external subs (and have them all in the same list). I will try this in the evening (don't know if it really is a good idea - let's see).

mcosmin222 commented:
I thought about that when i created the class, but I decided it is too complicated.

lukasf commented:
When I started moving the sync code into main SubtitleProvider class I realized we do not even need the ExternalSubsParser. See my latest changeset! I directly put the original SubtitleParser from external MSS into main MSS and things work very well.

mcosmin222 commented:
Cool. Nice catch. Just by eye-balling it, I think we need to do some detaching of references though, for example, the FrmpegReader is still referenced from the extractor MSS.

lukasf commented:
Good point. I added detach code and I can confirm that reader and external mss instances are destroyed as soon as the external sub is loaded.

mcosmin222 commented:
Looks good!

lukasf commented:
Are we good for merge, or shall we give this some more time for testing?

mcosmin222 commented:
I think we are good.
